src/PlayAudioUsingMediaStreams/Controllers/AnimalSoundboardController.cs

```diff
@@ -33,93 +40,108 @@ public async Task Get()
             HttpContext.Response.StatusCode = StatusCodes.Status400BadRequest;
         }
     }
-    
+
     private async Task Soundboard(WebSocket webSocket)
     {
+        string streamSid = null;
         var buffer = new byte[1024 * 4];
-        var receiveResult = await webSocket.ReceiveAsync(
-            new ArraySegment<byte>(buffer), CancellationToken.None);
+        var receiveResult = await webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);
+
+        await using var speechRecognitionStream = await _speechRecognitionService.InitStream();
+        while (!receiveResult.CloseStatus.HasValue &&
+               !_applicationLifetime.ApplicationStopping.IsCancellationRequested)
+        {
+            using var jsonDocument = JsonDocument.Parse(Encoding.UTF8.GetString(buffer, 0, receiveResult.Count));
+            var eventMessage = jsonDocument.RootElement.GetProperty("event").GetString();
+
+            switch (eventMessage)
+            {
+                case "connected":
+                    Console.WriteLine("Event: connected");
+                    break;
+                case "start":
+                    Console.WriteLine("Event: start");
+                    streamSid = jsonDocument.RootElement.GetProperty("streamSid").GetString();
+                    Console.WriteLine($"StreamId: {streamSid}");
 
-        var streamSid = string.Empty;
+                    // Do not await task, leave this task running in the background for the duration of the websocket connection
+                    var _ = ListenForSpeechRecognition(webSocket, streamSid, speechRecognitionStream)
+                        .ConfigureAwait(false);
+                    break;
+                case "media":
+                    var payload = jsonDocument.RootElement.GetProperty("media").GetProperty("payload").GetString();
+                    await _speechRecognitionService.SendAudio(payload);
+                    break;
+                case "stop":
+                    Console.WriteLine("Event: stop");
+                    break;
+            }
+
+            receiveResult = await webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);
+        }
+
+        if (receiveResult.CloseStatus.HasValue)
+        {
+            await webSocket.CloseAsync(
+                receiveResult.CloseStatus.Value,
+                receiveResult.CloseStatusDescription,
+                CancellationToken.None);
+        }
+        else if (_applicationLifetime.ApplicationStopping.IsCancellationRequested)
+        {
+            await webSocket.CloseAsync(
+                WebSocketCloseStatus.EndpointUnavailable,
+                "Server shutting down",
+                CancellationToken.None);
+        }
+    }
 
-        while (!receiveResult.CloseStatus.HasValue)
+    private async Task ListenForSpeechRecognition(
```

Comment:
I moved this code into its own method to reduce the amount of nesting, making it more readable.